File: app/src/main/java/com/boardgamegeek/ui/HotnessFragment.java
package com.boardgamegeek.ui;

import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Pair;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnLongClickListener;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.boardgamegeek.R;
import com.boardgamegeek.io.Adapter;
import com.boardgamegeek.io.BggService;
import com.boardgamegeek.model.HotGame;
import com.boardgamegeek.model.HotnessResponse;
import com.boardgamegeek.ui.loader.BggLoader;
import com.boardgamegeek.ui.loader.SafeResponse;
import com.boardgamegeek.util.ActivityUtils;
import com.boardgamegeek.util.AnimationUtils;
import com.boardgamegeek.util.ImageUtils;
import com.boardgamegeek.util.PreferencesUtils;
import com.boardgamegeek.util.PresentationUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import hugo.weaving.DebugLog;
import retrofit2.Call;

public class HotnessFragment extends Fragment implements LoaderManager.LoaderCallbacks<SafeResponse<HotnessResponse>>, ActionMode.Callback {
	private static final int LOADER_ID = 1;

	private HotGamesAdapter adapter;
	private ActionMode actionMode;
	private Unbinder unbinder;
	@BindView(R.id.root_container) CoordinatorLayout containerView;
	@BindView(android.R.id.progress) View progressView;
	@BindView(android.R.id.empty) TextView emptyView;
	@BindView(android.R.id.list) RecyclerView recyclerView;

	@DebugLog
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View rootView = inflater.inflate(R.layout.fragment_hotness, container, false);
		unbinder = ButterKnife.bind(this, rootView);
		setUpRecyclerView();
		return rootView;
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		getLoaderManager().initLoader(LOADER_ID, null, this);
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		if (unbinder != null) unbinder.unbind();
	}

	private void setUpRecyclerView() {
		recyclerView.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false));
		recyclerView.setHasFixedSize(true);
	}

	@Override
	public Loader<SafeResponse<HotnessResponse>> onCreateLoader(int id, Bundle data) {
		return new HotnessLoader(getActivity());
	}

	@Override
	public void onLoadFinished(Loader<SafeResponse<HotnessResponse>> loader, SafeResponse<HotnessResponse> data) {
		if (getActivity() == null) {
			return;
		}

		if (adapter == null) {
			adapter = new HotGamesAdapter(getActivity(),
				data == null || data.getBody() == null ? new ArrayList<HotGame>() : data.getBody().games,
				new Callback() {
					@Override
					public boolean onItemClick(int position) {
						if (actionMode == null) {
							return false;
						}
						toggleSelection(position);
						return true;
					}

					@Override
					public boolean onItemLongClick(int position) {
						if (actionMode != null) {
							return false;
						}
						actionMode = getActivity().startActionMode(HotnessFragment.this);
						toggleSelection(position);
						return true;
					}

					private void toggleSelection(int position) {
						adapter.toggleSelection(position);
						int count = adapter.getSelectedItemCount();
						if (count == 0) {
							actionMode.finish();
						} else {
							actionMode.setTitle(getResources().getQuantityString(R.plurals.msg_games_selected, count, count));
							actionMode.invalidate();
						}
					}
				});
			recyclerView.setAdapter(adapter);
		} else {
			adapter.notifyDataSetChanged();
		}

		if (data == null) {
			AnimationUtils.fadeOut(recyclerView);
			AnimationUtils.fadeIn(emptyView);
		} else if (data.hasError()) {
			emptyView.setText(data.getErrorMessage());
			AnimationUtils.fadeOut(recyclerView);
			AnimationUtils.fadeIn(emptyView);
		} else {
			AnimationUtils.fadeOut(emptyView);
			AnimationUtils.fadeIn(getActivity(), recyclerView, isResumed());
		}
		AnimationUtils.fadeOut(progressView);
	}

	@Override
	public void onLoaderReset(Loader<SafeResponse<HotnessResponse>> loader) {
	}

	private static class HotnessLoader extends BggLoader<SafeResponse<HotnessResponse>> {
		private final BggService bggService;

		public HotnessLoader(Context context) {
			super(context);
			bggService = Adapter.createForXml();
		}

		@Override
		public SafeResponse<HotnessResponse> loadInBackground() {
			Call<HotnessResponse> call = bggService.getHotness(BggService.HOTNESS_TYPE_BOARDGAME);
			return new SafeResponse<>(call);
		}
	}

	public interface Callback {
		boolean onItemClick(int position);

		boolean onItemLongClick(int position);
	}

	public class HotGamesAdapter extends RecyclerView.Adapter<HotGamesAdapter.ViewHolder> {
		private final LayoutInflater inflater;
		private final List<HotGame> games;
		private final Callback callback;
		private final SparseBooleanArray selectedItems;

		public HotGamesAdapter(Context context, List<HotGame> games, Callback callback) {
			this.games = games;
			this.callback = callback;
			inflater = LayoutInflater.from(context);
			selectedItems = new SparseBooleanArray();
			setHasStableIds(true);
		}

		@Override
		public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
			View view = inflater.inflate(R.layout.row_hotness, parent, false);
			return new ViewHolder(view);
		}

		@Override
		public void onBindViewHolder(ViewHolder holder, int position) {
			holder.bind(getItem(position), position);
		}

		@Override
		public int getItemCount() {
			return games == null ? 0 : games.size();
		}

		@Override
		public long getItemId(int position) {
			return position;
		}

		public HotGame getItem(int position) {
			return games.get(position);
		}

		public class ViewHolder extends RecyclerView.ViewHolder {
			private int gameId;
			private String gameName;
			@BindView(R.id.name) TextView name;
			@BindView(R.id.year) TextView year;
			@BindView(R.id.rank) TextView rank;
			@BindView(R.id.thumbnail) ImageView thumbnail;

			public ViewHolder(View itemView) {
				super(itemView);
				ButterKnife.bind(this, itemView);
			}

			public void bind(HotGame game, final int position) {
				if (game == null) return;
				gameId = game.id;
				gameName = game.name;
				name.setText(game.name);
				year.setText(PresentationUtils.describeYear(name.getContext(), game.yearPublished));
				rank.setText(String.valueOf(game.rank));
				ImageUtils.loadThumbnail(game.thumbnailUrl, thumbnail);

				itemView.setActivated(selectedItems.get(position, false));

				itemView.setOnClickListener(new OnClickListener() {
					@Override
					public void onClick(View v) {
						boolean handled = false;
						if (callback != null) {
							handled = callback.onItemClick(position);
						}
						if (!handled) {
							ActivityUtils.launchGame(itemView.getContext(), gameId, gameName);
						}
					}
				});

				itemView.setOnLongClickListener(new OnLongClickListener() {
					@Override
					public boolean onLongClick(View v) {
						return callback != null && callback.onItemLongClick(position);
					}
				});
			}
		}

		public void toggleSelection(int position) {
			if (selectedItems.get(position, false)) {
				selectedItems.delete(position);
			} else {
				selectedItems.put(position, true);
			}
			notifyItemChanged(position);
		}

		public void clearSelections() {
			selectedItems.clear();
			notifyDataSetChanged();
		}

		public int getSelectedItemCount() {
			return selectedItems.size();
		}

		public List<Integer> getSelectedItems() {
			List<Integer> items = new ArrayList<>(selectedItems.size());
			for (int i = 0; i < selectedItems.size(); i++) {
				items.add(selectedItems.keyAt(i));
			}
			return items;
		}
	}

	@Override
	public boolean onCreateActionMode(ActionMode mode, Menu menu) {
		adapter.clearSelections();
		MenuInflater inflater = mode.getMenuInflater();
		inflater.inflate(R.menu.game_context, menu);
		return true;
	}

	@Override
	public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
		int count = adapter.getSelectedItemCount();
		menu.findItem(R.id.menu_log_play).setVisible(count == 1 && PreferencesUtils.showLogPlay(getActivity()));
		menu.findItem(R.id.menu_log_play_quick).setVisible(PreferencesUtils.showQuickLogPlay(getActivity()));
		menu.findItem(R.id.menu_link).setVisible(count == 1);
		return true;
	}

	@Override
	public void onDestroyActionMode(ActionMode mode) {
		actionMode = null;
		adapter.clearSelections();
	}

	@Override
	public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
		HotGame game = adapter.getItem(adapter.getSelectedItems().get(0));
		switch (item.getItemId()) {
			case R.id.menu_log_play:
				mode.finish();
				ActivityUtils.logPlay(getActivity(), game.id, game.name, game.thumbnailUrl, game.thumbnailUrl, false);
				return true;
			case R.id.menu_log_play_quick:
				mode.finish();
				String text = getResources().getQuantityString(R.plurals.msg_logging_plays, adapter.getSelectedItemCount());
				Snackbar.make(containerView, text, Snackbar.LENGTH_SHORT).show();
				for (int position : adapter.getSelectedItems()) {
					HotGame g = adapter.getItem(position);
					ActivityUtils.logQuickPlay(getActivity(), g.id, g.name);
				}
				return true;
			case R.id.menu_share:
				mode.finish();
				if (adapter.getSelectedItemCount() == 1) {
					ActivityUtils.shareGame(getActivity(), game.id, game.name);
				} else {
					List<Pair<Integer, String>> games = new ArrayList<>(adapter.getSelectedItemCount());
					for (int position : adapter.getSelectedItems()) {
						HotGame g = adapter.getItem(position);
						games.add(Pair.create(g.id, g.name));
					}
					ActivityUtils.shareGames(getActivity(), games);
				}
				return true;
			case R.id.menu_link:
				mode.finish();
				ActivityUtils.linkBgg(getActivity(), game.id);
				return true;
		}
		return false;
	}
}

File: app/src/main/java/com/boardgamegeek/ui/HotnessFragment.kt
package com.boardgamegeek.ui

import android.os.Bundle
import android.util.SparseBooleanArray
import android.view.*
import androidx.compose.foundation.background
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.CalendarToday
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.tooling.preview.PreviewParameter
import androidx.compose.ui.tooling.preview.PreviewParameterProvider
import androidx.compose.ui.unit.dp
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.RecyclerView
import com.boardgamegeek.R
import com.boardgamegeek.auth.Authenticator
import com.boardgamegeek.databinding.FragmentHotnessBinding
import com.boardgamegeek.extensions.*
import com.boardgamegeek.model.HotGame
import com.boardgamegeek.model.Status
import com.boardgamegeek.ui.adapter.AutoUpdatableAdapter
import com.boardgamegeek.ui.compose.*
import com.boardgamegeek.ui.theme.BggAppTheme
import com.boardgamegeek.ui.viewmodel.HotnessViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlin.properties.Delegates

@AndroidEntryPoint
class HotnessFragment : Fragment(), ActionMode.Callback {
    private var _binding: FragmentHotnessBinding? = null
    private val binding get() = _binding!!
    private val viewModel by activityViewModels<HotnessViewModel>()
    private val adapter: HotGamesAdapter by lazy { createAdapter() }
    private var actionMode: ActionMode? = null

    @Suppress("RedundantNullableReturnType")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentHotnessBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.recyclerView.setHasFixedSize(true)
        binding.recyclerView.adapter = adapter

        viewModel.errorMessage.observe(viewLifecycleOwner) { event ->
            event.getContentIfNotHandled()?.let {
                binding.coordinatorLayout.snackbar(it)
            }
        }

        viewModel.loggedPlayResult.observe(viewLifecycleOwner) { event ->
            event.getContentIfNotHandled()?.let {
                requireContext().notifyLoggedPlay(it)
            }
        }

        viewModel.hotGames.observe(viewLifecycleOwner) {
            it?.let { (status, data, message) ->
                when (status) {
                    Status.REFRESHING -> binding.progressView.show()
                    Status.ERROR -> {
                        binding.emptyView.text = message
                        binding.emptyView.isVisible = true
                        binding.recyclerView.isVisible = false
                        binding.progressView.hide()
                    }
                    Status.SUCCESS -> {
                        val games = data.orEmpty()
                        adapter.games = games
                        if (games.isEmpty()) {
                            binding.emptyView.setText(R.string.empty_hotness)
                            binding.emptyView.isVisible = true
                            binding.recyclerView.isVisible = false
                        } else {
                            binding.emptyView.isVisible = false
                            binding.recyclerView.isVisible = true
                        }
                        binding.progressView.hide()
                    }
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding.recyclerView.adapter = null
        _binding = null
    }

    private fun createAdapter(): HotGamesAdapter {
        return HotGamesAdapter(object : Callback {
            override fun onItemClick(position: Int): Boolean {
                if (actionMode == null) return false
                toggleSelection(position)
                return true
            }

            override fun onItemLongClick(position: Int): Boolean {
                if (actionMode != null) return false
                actionMode = requireActivity().startActionMode(this@HotnessFragment)
                if (actionMode == null) return false
                toggleSelection(position)
                return true
            }

            private fun toggleSelection(position: Int) {
                adapter.toggleSelection(position)
                actionMode?.let {
                    if (adapter.selectedItemCount == 0) {
                        it.finish()
                    } else {
                        it.invalidate()
                    }
                }
            }
        })
    }

    interface Callback {
        fun onItemClick(position: Int): Boolean
        fun onItemLongClick(position: Int): Boolean
    }

    inner class HotGamesAdapter(private val callback: Callback?) : RecyclerView.Adapter<HotGamesAdapter.ViewHolder>(), AutoUpdatableAdapter {
        private val selectedItems = SparseBooleanArray()

        init {
            setHasStableIds(true)
        }

        var games: List<HotGame> by Delegates.observable(emptyList()) { _, old, new ->
            autoNotify(old, new) { o, n -> o == n }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ViewHolder(ComposeView(parent.context))

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.bind(position)
        }

        override fun getItemCount() = games.size

        override fun getItemId(position: Int) = games.getOrNull(position)?.id?.toLong() ?: RecyclerView.NO_ID

        inner class ViewHolder(private val composeView: ComposeView) : RecyclerView.ViewHolder(composeView) {
            fun bind(position: Int) {
                composeView.setContent {
                    HotnessListItem(
                        hotGame = games[position],
                        isSelected = selectedItems[position, false],
                        onClick = { hotGame ->
                            if (callback?.onItemClick(position) != true) {
                                GameActivity.start(requireContext(), hotGame.id, hotGame.name, hotGame.thumbnailUrl)
                            }
                        },
                        onLongClick = {
                            callback?.onItemLongClick(position)
                        }
                    )
                }
            }
        }

        fun toggleSelection(position: Int) {
            selectedItems.toggle(position)
            notifyItemChanged(position)
        }

        fun clearSelections() {
            val oldSelectedItems = selectedItems.clone()
            selectedItems.clear()
            oldSelectedItems.filterTrue().forEach { notifyItemChanged(it) }
        }

        val selectedItemCount: Int
            get() = selectedItems.filterTrue().size

        fun getSelectedGames() = selectedItems.filterTrue().mapNotNull { games.getOrNull(it) }
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        adapter.clearSelections()
        mode.menuInflater.inflate(R.menu.game_context, menu)
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        val count = adapter.selectedItemCount
        mode.title = resources.getQuantityString(R.plurals.msg_games_selected, count, count)
        if (Authenticator.isSignedIn(context)) {
            menu.findItem(R.id.menu_log_play_form).isVisible = count == 1
            menu.findItem(R.id.menu_log_play_wizard).isVisible = count == 1
            menu.findItem(R.id.menu_log_play).isVisible = true
        } else {
            menu.findItem(R.id.menu_log_play).isVisible = false
        }
        menu.findItem(R.id.menu_link).isVisible = count == 1
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        actionMode = null
        adapter.clearSelections()
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        val selectedGames = adapter.getSelectedGames()
        if (selectedGames.isEmpty()) return false
        when (item.itemId) {
            R.id.menu_log_play_form -> {
                selectedGames.firstOrNull()?.let { game ->
                    LogPlayActivity.logPlay(requireContext(), game.id, game.name, game.thumbnailUrl)
                }
            }
            R.id.menu_log_play_quick -> {
                binding.coordinatorLayout.snackbar(resources.getQuantityString(R.plurals.msg_logging_plays, adapter.selectedItemCount))
                for (game in selectedGames) {
                    viewModel.logQuickPlay(game.id, game.name)
                }
            }
            R.id.menu_log_play_wizard -> {
                selectedGames.firstOrNull()?.let { game ->
                    NewPlayActivity.start(requireContext(), game.id, game.name)
                }
            }
            R.id.menu_share -> {
                val shareMethod = "Hotness"
                if (selectedGames.size == 1) {
                    selectedGames.firstOrNull()?.let { game ->
                        requireActivity().shareGame(game.id, game.name, shareMethod)
                    }
                } else {
                    requireActivity().shareGames(selectedGames.map { it.id to it.name }, shareMethod)
                }
            }
            R.id.menu_link -> {
                selectedGames.firstOrNull()?.let { game ->
                    context.linkBgg(game.id)
                }
            }
            else -> return false
        }
        mode.finish()
        return true
    }
}

@Composable
fun HotnessListItem(
    hotGame: HotGame,
    modifier: Modifier = Modifier,
    onClick: (hotGame: HotGame) -> Unit = {},
    onLongClick: () -> Unit = {},
    isSelected: Boolean = false,
) {
    Row(
        horizontalArrangement = Arrangement.Start,
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier // TODO move some of this to other modifier?
            .fillMaxWidth()
            .background(if (isSelected) MaterialTheme.colorScheme.primaryContainer else MaterialTheme.colorScheme.surface)
            .combinedClickable(
                onLongClick = onLongClick,
            ) {
                onClick(hotGame)
            }
            .padding(
                horizontal = dimensionResource(R.dimen.material_margin_horizontal),
                vertical = 12.dp,
            )
            .then(modifier)
    ) {
        ListItemIndex(hotGame.rank)
        ListItemThumbnail(hotGame.thumbnailUrl)
        Column {
            ListItemPrimaryText(hotGame.name)
            ListItemSecondaryText(
                hotGame.yearPublished.asYear(LocalContext.current),
                modifier = modifier.padding(bottom = ListItemTokens.verticalTextPadding),
                icon = Icons.Outlined.CalendarToday,
            )
        }
    }
}

@PreviewLightDark
@Composable
private fun HotnessListItemPreview(
    @PreviewParameter(HotGamePreviewParameterProvider::class) hotGame: HotGame,
) {
    BggAppTheme {
        HotnessListItem(hotGame, Modifier)
    }
}

private class HotGamePreviewParameterProvider : PreviewParameterProvider<HotGame> {
    override val values = sequenceOf(
        HotGame(
            rank = 1,
            id = 99,
            name = "Spirit Island",
            yearPublished = 2019,
        ),
        HotGame(
            rank = 22,
            id = 99,
            name = "Star Wars: the Deck Building Game",
            yearPublished = 2023,
        ),
        HotGame(
            rank = 50,
            id = 99,
            name = "Sky Team",
            yearPublished = 2022,
        )
    )
}
